### ISSUE TITLE
Pro sees listing requirements list

### ISSUE THREAD

**Issue by situatedbit** [loconomics/loconomics#433]
Profile: [Devin, aspirational dog walker](https://github.com/loconomics/loconomics/blob/master/docs/Personas/Devin.md)

Story: as a new professional who moved quickly through onboarding (skipping some steps), I am eager to make my listing public. However, I need to know which steps I need to finish before I can make my listing available. On the listing admin page, I want to see a list of steps required for me to make my profile visible and because I don’t know my way around the app, I want that list to include buttons or links to the forms where I complete those steps. (This includes steps that are not specific to a listing, such as scheduling preferences, payment details, etc).

Background:
This story has value independent of onboarding. Currently we tell users when their listings are incomplete, but we don't actually tell them what they need to complete!

Additionally, once this item is in place we can start allowing skipping of required items during onboarding.

See #170 for possibly more detail.

**Comment by situatedbit** [comment]
Here's my suggestion for how the list and the navigation links work:

![listing](https://cloud.githubusercontent.com/assets/6644016/25875213/109ea3c4-34cb-11e7-9e29-495b60741a03.jpg)

(https://loconomics.mybalsamiq.com/projects/listing-requirements/listing-requirements)

**Comment by situatedbit** [comment]
I assume we're using the Alerts table for this text. There are two columns for each alert: headline display and text display. I'm not sure what the difference is, but so far I only see the need for 1 string per requirement. This is what I think the copy should look like:

AlertName | AlertHeadlineDisplay/TextDisplay
--- | ---
pricingdetails | How much do your services cost? communicate accurate pricing
availability | When are you available to work? |
personalinfo | How do we reach you? |
payment | How do you want to be paid? |
positionservices | What services do you provide? |
location | Where can you work? |
required-backgroundcheck | Undergo a background check |  
required-professionallicense | Verify your professional license or certification |

**Comment by situatedbit** [comment]
- server-side model
- api call
- client-side model
- client-side API interface
- view model
- integration with marketplaceProfile activity

**Comment by situatedbit** [comment]
Time-boxed for 2hrs of research for Matt to make a proposal

**Comment by situatedbit** [comment]
@IagoSRL @joshdanielson Can you all provide feedback on this? After that we can find a way to point it so I can get working on it.

Here is my proposal for the API, database, and client-side model.
1. we should remove as many unused alerts database objects from the database as is feasible
2. we should move the text content from the database to the codebase to make it easier to manage updates (my sense is we _rarely_ change this text anyway)

# Database Changes

Drop table ``AlertType``. We don't use it. If we need to add something similar back in the future, then we can do it.

Keep the following fields for the ``Alerts`` table:
- AlertID
- AlertName
- CreatedDate
- UpdatedDate
- ModifiedBy
- Active
- Required
- PositionSpecific
- DisplayRank
- bookMeButtonRequired

Keep the following fields for the ``UserAlerts`` table:
- UserID
- PositionID
- AlertID
- CreatedDate
- UpdatedDate
- ModifiedBy
- Active (though, not sure if this is currently used)

# API

```
me/profile-alerts/{jobTitleID} => returns array of alerts objects
```

# Client-Side Models

## New: appmodel/AppModel.profileAlert

```javascript
get(positionID) => array of active alerts
```

## New: models/ProfileAlert

Fields: 
- alertID
- name
- description => derived in code from name
- isRequired
- isPositionSpecific

## models/UserJobTitle
New Fields:
- alerts: { isArray: true, Model: ProfileAlert }

New Function:
```javascript
requiredAlerts => array of ProfileAlert objects where isRequired is true
```

**Comment by joshdanielson** [comment]
@situatedbit seems pretty thorough to me. From what I see, there will be no content, e.g. "When are you available to work?" generated from the database going forward. I agree just clarifying.

**Comment by IagoSRL** [comment]
About the table fields clean-up:
- `Alerts` table: the ProviderAlert and CustomerAlert fields are in use at `[fxCheckAlertAffectsUser]` (SQL UDF) used in all `TestAlert*` procs.
  - This is because a data change in the API/backend would call the related `TestAlert*` for any user, and that would check if the user is affected by the alert; there are different requirements for clients and professionals.
  - We can rename the columns to the new names, if that's confusing.
- `UserAlert` table:
  - The `Active` field is not in use (rather than to set the value to 1 when inserting a record), I'm OK with remove it since now, when user completes an alert, the record is just removed rather than set this to 0.
    - @joshdanielson If there is any reason to keep the record, like to track when was completed (there is an unused `CompletedDate` field), we can keep it, but I think that's for another issue asking to implement that feature.
  - The `Dismissed` field: was in use (dismissable alerts was implemented in release-6.3 site) but not sure if we want to keep the feature (needs still implementation in the App and REST API, and only available to optional alerts); we can remove it now and open another issue if wanted. This would be blocked by implementing listing of optional alerts, here we are targeting only the required ones.

I agree with the unused alerts clean-up, but take care that requires touch some code (not only data), like the `TestAlert*` procs and code where is executed (or errors will start being triggered); maybe not a top priority.
If is expected some of that alerts may be recovered at some point, we can open an issue for each one with a copy there of the data record and point to the commit that removed affected code including copy of SQL procs/UDFs (for easy recovering).

-------------------

@situatedbit Since you want to add an `alerts` property at models/UserJobTitle, you may want to include that data as part of the REST API response to GET `/me/user-job-profile/{jobTitleID}`, to be even easier to access it (rather than run a parallel request to fill in the data). I have no preference, though.

**Comment by joshdanielson** [comment]
@IagoSRL I don't see us needing to know the completedDate at this point, just whether or not it's been completed.

**Comment by IagoSRL** [comment]
@joshdanielson OK, thanks.
About the dismiss feature, if that's wanted for optional/enhancements, please open an issue (blocked/part of #412). I agree to remove the field for now though, and recreate at that scope.

**Comment by situatedbit** [comment]
@IagoSRL I updated my proposal to keep ProviderAlert and CustomerAlert, and removing Active and Dismissed from UserAlerts.

I agree with your suggestion to add the alerts to the existing user job profile api call. I'll change that in the proposal as well.

**Comment by joshdanielson** [comment]
@situatedbit I think we can figure out a better system for highlighting ways to enhance their listings that don't require an "alert" they have to dismiss, don't you agree?

**Comment by situatedbit** [comment]
@joshdanielson yes, but the names we use in the database don't have to dictate how they're displayed to the user. So we can still use ''alerts" but render them in different ways.

**Comment by situatedbit** [comment]
@IagoSRL because this item will require changes to stored procedures, I'm going to move #455 (adding stored procedures and user defined functions to the repo) ahead of this in priority. 

Also, I'd like to estimate this item. I already know what I'm going to estimate it as. Iago, do you feel comfortable adding your estimate? (Or, we could do it in Slack whenever you want)

**Comment by situatedbit** [comment]
Status: I'm ready to clean up the database tables.
